## Q1

In [1]:
import numpy as np
import pandas as pd

In [2]:
nationality=pd.DataFrame(["British","American","Spanish","Italian","French","Turkish"]) #nominal
timeofday=pd.DataFrame(["morning","afternoon","night","evening","morning","evening"]) #ordinal
bdate=pd.DataFrame([2009,2016,2005,2013,2002,2011]) #interval
weight=pd.DataFrame([50,68,45,47,56,69]) #ratio
dfs=[nationality,timeofday,bdate,weight]
ABCD=pd.concat(dfs,axis=1)
ABCD.columns=["nationality","timeofday","bdate","weight"]
ABCD.head()

nationality  timeofday  bdate  weight
0     British    morning   2009      50
1    American  afternoon   2016      68
2     Spanish      night   2005      45
3     Italian    evening   2013      47
4      French    morning   2002      56

## Q2

In [3]:
data=pd.read_csv("data.csv")
data.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [4]:
data.describe().T

count          mean           std          min  \
id                       569.0  3.037183e+07  1.250206e+08  8670.000000   
radius_mean              569.0  1.412729e+01  3.524049e+00     6.981000   
texture_mean             569.0  1.928965e+01  4.301036e+00     9.710000   
perimeter_mean           569.0  9.196903e+01  2.429898e+01    43.790000   
area_mean                569.0  6.548891e+02  3.519141e+02   143.500000   
smoothness_mean          569.0  9.636028e-02  1.406413e-02     0.052630   
compactness_mean         569.0  1.043410e-01  5.281276e-02     0.019380   
concavity_mean           569.0  8.879932e-02  7.971981e-02     0.000000   
concave points_mean      569.0  4.891915e-02  3.880284e-02     0.000000   
symmetry_mean            569.0  1.811619e-01  2.741428e-02     0.106000   
fractal_dimension_mean   569.0  6.279761e-02  7.060363e-03     0.049960   
radius_se                569.0  4.051721e-01  2.773127e-01     0.111500   
texture_se               569.0  1.216853e+00  5.516484e-01     0.360200   
perimeter_se             569.0  2.866059e+00  2.021855e+00     0.757000   
area_se                  569.0  4.033708e+01  4.549101e+01     6.802000   
smoothness_se            569.0  7.040979e-03  3.002518e-03     0.001713   
compactness_se           569.0  2.547814e-02  1.790818e-02     0.002252   
concavity_se             569.0  3.189372e-02  3.018606e-02     0.000000   
concave points_se        569.0  1.179614e-02  6.170285e-03     0.000000   
symmetry_se              569.0  2.054230e-02  8.266372e-03     0.007882   
fractal_dimension_se     569.0  3.794904e-03  2.646071e-03     0.000895   
radius_worst             569.0  1.626919e+01  4.833242e+00     7.930000   
texture_worst            569.0  2.567722e+01  6.146258e+00    12.020000   
perimeter_worst          569.0  1.072612e+02  3.360254e+01    50.410000   
area_worst               569.0  8.805831e+02  5.693570e+02   185.200000   
smoothness_worst         569.0  1.323686e-01  2.283243e-02     0.071170   
compactness_worst        569.0  2.542650e-01  1.573365e-01     0.027290   
concavity_worst          569.0  2.721885e-01  2.086243e-01     0.000000   
concave points_worst     569.0  1.146062e-01  6.573234e-02     0.000000   
symmetry_worst           569.0  2.900756e-01  6.186747e-02     0.156500   
fractal_dimension_worst  569.0  8.394582e-02  1.806127e-02     0.055040   
Unnamed: 32                0.0           NaN           NaN          NaN   

                                   25%            50%           75%  \
id                       869218.000000  906024.000000  8.813129e+06   
radius_mean                  11.700000      13.370000  1.578000e+01   
texture_mean                 16.170000      18.840000  2.180000e+01   
perimeter_mean               75.170000      86.240000  1.041000e+02   
area_mean                   420.300000     551.100000  7.827000e+02   
smoothness_mean               0.086370       0.095870  1.053000e-01   
compactness_mean              0.064920       0.092630  1.304000e-01   
concavity_mean                0.029560       0.061540  1.307000e-01   
concave points_mean           0.020310       0.033500  7.400000e-02   
symmetry_mean                 0.161900       0.179200  1.957000e-01   
fractal_dimension_mean        0.057700       0.061540  6.612000e-02   
radius_se                     0.232400       0.324200  4.789000e-01   
texture_se                    0.833900       1.108000  1.474000e+00   
perimeter_se                  1.606000       2.287000  3.357000e+00   
area_se                      17.850000      24.530000  4.519000e+01   
smoothness_se                 0.005169       0.006380  8.146000e-03   
compactness_se                0.013080       0.020450  3.245000e-02   
concavity_se                  0.015090       0.025890  4.205000e-02   
concave points_se             0.007638       0.010930  1.471000e-02   
symmetry_se                   0.015160       0.018730  2.348000e-02   
fractal_dimension_se          0.002248       0.003187  4.5

In [5]:
data.describe()[["radius_mean", "area_mean", "compactness_mean"]].T

count        mean         std        min        25%  \
radius_mean       569.0   14.127292    3.524049    6.98100   11.70000   
area_mean         569.0  654.889104  351.914129  143.50000  420.30000   
compactness_mean  569.0    0.104341    0.052813    0.01938    0.06492   

                        50%       75%        max  
radius_mean        13.37000   15.7800    28.1100  
area_mean         551.10000  782.7000  2501.0000  
compactness_mean    0.09263    0.1304     0.3454

Bu tabloda, öncelikle yarıçap ortalamasında 569 veri üzerinde çalışıldığı ve ortalama değerinin 14.127292 olduğu görülmektedir. Standart sapması ise 3.524049'dur. Bu durum ortalamadan olan sapmaları ifade etmektedir. Bu değerin küçük olması nedeniyle değerlerin ortalama etrafında olduğu söylenebilir. Değerler 6.98100 ile 28.1100 arasında değişmektedir. Değişim aralığının küçük olması nedeniyle değerlerin birbirine yakın olduğu söylenebilir. Aynı zamanda medyanının 13.37000 olması ve ortalama ile yakın olması aykırılıkların azlığını ifade etmektedir. 

Alan ortalamasında da 569 veri üzerinde çalışılmıştır ve ortalama değerinin 654.889104 olduğu görülmektedir. Standart sapmasının 351.914129 olması değerlere göre yorumlandığında büyük bir değer olmadığı söylenebilir. Değerler 143.50000 ile 2501.0000 arasında değişmektedir dolayısıyla değişim aralığına bakıldığında değerlerin birbirinden uzak olduğu yorumu yapılabilir. Medyanın 551.10000 olması ve ortalama ile yakın değerler olması yarıçap ortalamasında olduğu gibi aykırılıkların azlığını ifade ermektedir.

Yoğunluk ortalamasında ise diğerlerinde olduğu gibi 569 veri üzerinde çalışılmıştır. Burada öncelikle değerlerin çok küçük olduğunu ve ortalamasının 0.104341 olduğu görülmektedir. Standart sapmasının 0.052813 olduğu görülmektedir. Küçük bir değer olduğundan verilerin ortalama etrafınnda değiştiği söylenebilir. Aynı zamanda medyanın 0.09263 olması ve ortalama ile çok yakın olması aykırılıkların azlığını ifade etmektedir. Değerler 0.01938 ile 0.3454 arasında değişmektedir. Değişim aralığına bakıldığında değerlerin birbirinden uzak olduğu yorumu yapılabilir.

In [6]:
Q1=data.quantile(0.25)
Q3=data.quantile(0.75)

In [7]:
IQR=Q3-Q1

In [8]:
lower_outlier=Q1-1.5*IQR
higher_outlier=Q3+1.5*IQR

In [9]:
lower_outlier[["radius_mean", "area_mean", "compactness_mean"]]

radius_mean           5.5800
area_mean          -123.3000
compactness_mean     -0.0333
dtype: float64

In [10]:
higher_outlier[["radius_mean", "area_mean", "compactness_mean"]]

radius_mean           21.90000
area_mean           1326.30000
compactness_mean       0.22862
dtype: float64

In [11]:
o1=(data["radius_mean"]<lower_outlier["radius_mean"]) | (data["radius_mean"]>higher_outlier["radius_mean"])
data[o1]["radius_mean"]

82     25.22
108    22.27
122    24.25
164    23.27
180    27.22
202    23.29
212    28.11
236    23.21
339    23.51
352    25.73
369    22.01
461    27.42
503    23.09
521    24.63
Name: radius_mean, dtype: float64

Yarıçap ortalaması hakkında 5.5800'den küçük ve 21.90000'dan büyük veriler için aykırı değerler tanımı yapılmıştır. Yukarıda görüldüğü üzere 14 tane aykırı değer bulunmaktadır.

In [12]:
o2=(data["area_mean"]<lower_outlier["area_mean"]) | (data["area_mean"]>higher_outlier["area_mean"])
data[o2]["area_mean"]

23     1404.0
82     1878.0
108    1509.0
122    1761.0
164    1686.0
180    2250.0
202    1685.0
212    2499.0
236    1670.0
250    1364.0
265    1419.0
272    1491.0
339    1747.0
352    2010.0
368    1546.0
369    1482.0
372    1386.0
373    1335.0
393    1407.0
449    1384.0
461    2501.0
503    1682.0
521    1841.0
563    1347.0
564    1479.0
Name: area_mean, dtype: float64

Alan ortalaması hakkında -123.3000'ten küçük ve 1326.30000'ten büyük veriler için aykırı değerler tanımı yapılmıştır. Yukarıda görüldüğü üzere 25 tane aykırı değer bulunmaktadır.

In [13]:
o3=(data["compactness_mean"]<lower_outlier["compactness_mean"]) | (data["compactness_mean"]>higher_outlier["compactness_mean"])
data[o3]["compactness_mean"]

0      0.2776
3      0.2839
9      0.2396
12     0.2458
14     0.2293
78     0.3454
82     0.2665
108    0.2768
122    0.2867
181    0.2832
190    0.2413
258    0.3114
351    0.2364
352    0.2363
400    0.2576
567    0.2770
Name: compactness_mean, dtype: float64

Yoğunluk ortalaması hakkında -0.0333'ten küçük ve 0.22862'den büyük veriler için aykırı değerler tanımı yapılmıştır. Yukarıda görüldüğü üzere 16 tane aykırı değer bulunmaktadır.

## Q3

In [14]:
data[["radius_mean", "area_mean", "compactness_mean"]].corr().T

radius_mean  area_mean  compactness_mean
radius_mean          1.000000   0.987357          0.506124
area_mean            0.987357   1.000000          0.498502
compactness_mean     0.506124   0.498502          1.000000

Korelasyon analizi, değişkenler arasındaki ilişkiyi, bu ilişkinin yönü ve şiddetiyle ilgili bilgiler sağlayan istatistiksel bir yöntemdir. Burada yarıçap, alan ve yoğunluk ortalamalarının korelasyon değerleri bulunmuştur. 

Bu değerler şu şekilde yorumlanabilir: 

Alan ortalaması ve yarıçap ortalaması arasındaki ilişkiyi 0.987357 değeri ifade etmektedir. Bu değer mükemmel pozitif korelasyona çok yakın bir değerdir. Yoğunluk ortalaması ve yarıçap arasındaki ilişkiyi 0.506124 değeri gösterir. Bu değer düşük pozitif korelasyonu ifade eder. Yoğunluk ve alan ortalaması arasındaki ilişkiyi ifade eden 0.498502 değeri ise yine aynı şekilde düşük pozitif korelasyona yakın bir değerdir.

In [15]:
data[["radius_mean", "area_mean", "compactness_mean"]].cov().T

radius_mean      area_mean  compactness_mean
radius_mean         12.418920    1224.483409          0.094197
area_mean         1224.483409  123843.554318          9.264931
compactness_mean     0.094197       9.264931          0.002789

Kovaryans, iki değişken arasındaki ilişkinin değişkenlik ölçüsüdür. Bu tabloyla birlikte hepsinin birbiri arasında pozitif bir ilişki olduğu görülmektedir.

In [16]:
from scipy.stats import pearsonr
x = pearsonr(data["radius_mean"],data["area_mean"])
y = pearsonr(data["radius_mean"],data["compactness_mean"])
z = pearsonr(data["area_mean"],data["compactness_mean"])

print(f"{x}\n")
print(f"{y}\n")
print(f"{z}\n")

(0.9873571700566124, 0.0)

(0.5061235775263635, 2.4142374076952573e-38)

(0.4985016821524118, 4.469262412798543e-37)



Bu yöntemde korelasyonun anlamlılığı test edilir. Örneğin yarıçap ortalaması ile alan ortalaması arasındaki ilişki için p-value değeri 0'dır. Dolayısıyla hipotez reddedilir. Değişkenler arasında anlamlı bir korelasyonun olup olmadığı sonucuna varılmış olunur.

In [17]:
from scipy.stats import spearmanr
x = spearmanr(data["radius_mean"],data["area_mean"])
y = spearmanr(data["radius_mean"],data["compactness_mean"])
z = spearmanr(data["area_mean"],data["compactness_mean"])

print(f"{x}\n")
print(f"{y}\n")
print(f"{z}\n")

SpearmanrResult(correlation=0.9996020278067481, pvalue=0.0)

SpearmanrResult(correlation=0.49757760995795647, pvalue=6.3349241442214955e-37)

SpearmanrResult(correlation=0.4889883974982949, pvalue=1.5419515156000542e-35)



Bir hipotez testinde varsayımlar sağlanmadığında ise Spearman yöntemi kullanılır. Örneğin burada yarıçap ortalaması ile yoğunluk ortalaması arasındaki anlamlılık test edilmek istenildiğinde p-value değerine bakılır. Burada p-value değeri çok küçük bir değerdir. Dolayısıyla hipotez reddedilir. 

In [18]:
from numpy import mean, std 
from math import sqrt
import sys

def cohen_d(x,y):
        return (mean(x) - mean(y)) / sqrt((std(x, ddof=1) ** 2 + std(y, ddof=1) ** 2) / 2.0)

c0 = data["radius_mean"]
c1 = data["area_mean"]
c2 = data["compactness_mean"]
print(cohen_d(c1,c0))
print(cohen_d(c0,c2))
print(cohen_d(c1,c2))


2.5748571535813967
5.62682911222856
2.631339308769588


Etki büyüklüğünün hesaplanmasında en yaygın kullanılanı Cohen yöntemidir. Bu yöntem tarafından geliştirilen hesaplama d'dir. Cohen genel bir öneri olmak üzere, d değerinin 0,2’den küçük olması durumunda, etki büyüklüğünün zayıf, 0.5 olması durumunda orta ve 0,8’den büyük olması durumunda ise kuvvetli olarak tanımlanabileceğini söylemektedir. Burada ise yarıçap, alan ve yoğunluk ortalamaları karşılaştırılmıştır. Hepsinin sonucu 0,8'den büyük olduğundan etki büyüklükleri kuvvetli olarak yorumlanabilir.

## Q4

### Bayes Teoremi Örnekleri

1- PCR testi olan insanlarının %60'ının gerçekten hasta olduğunu ve belirli bir markanın isabet oranlarının test edilmek istenildiğini varsayalım. Test yaptıran kişi gerçekten hastaysa %95 ihtimalle testin pozitif sonuç verdiği ve test yaptıran kişi hasta değilse %85 oranında negatif sonuç verdiğini düşünelim. Bu durumda test olan bir kişinin sonucu pozitif çıkmışsa gerçekten hasta olma olasılığını bulmak için Bayes Teoremi kullanılabilir.

2- Bir şirket üyelerinin 1. işletmeden %50, 2. işletmeden %30 ve 3. işletmeden %15 olacak şekilde 3 farklı şirketten araba kiraladıklarını varsayalım. 1. işletmeden gelen arabaların %8'i, 2. işletmeden gelen arabaların %10'u ve 3. işletmeden gelen arabaların ise %20'si bakım gerektiriyorsa bu durumda kiralanan arabanın örneğin 1. işletmeden gelme olasılığı Bayes Teoremi ile bulunabilir.

3- Bir evde alarm sistemi olduğunu varsayalım. Bu alarm sisteminin tehlike olduğu zaman alarm verme olasılığı %99, tehlike olmadığında ise %0.2'dir. Eğer bu alarm çalışırsa ve alarmın tehlike nedeniyle çalmış olma olasılığını hesaplamak istersek Bayes Teoremi kullanılabilir.

4- Bir fabrikada üç farklı makinede üretim yapıldığını ve üretilen ürünlerin %35'inin 1. makineden, %45'inin 2. makineden ve %20'sinin ise 3. makineden üretildiğini varsayalım. 1. makinedeki üretimlerin %4'ü, 2. makinedeki üretimlerin %5'i, 3. makinedeki üretimlerin ise %7'sinin hatalı olduğu biliniyor. Bu durumda rastgele alınan bir ürünün hatalı olduğu bilindiğine göre bu ürünün 1. makineden üretilmiş olma olasılığını Bayes Teoremi ile bulabiliriz.

5- Bir sınıftaki öğrencilerin %60'ı sayısal, %30'u eşit ağırlık, %10'unun ise sözel seçtiğini varsayalım. Öğrencilerin ingilizceye olan ilgisini ölçmek için yapılan bir ankette sayısal seçen öğrencilerin %35'i, eşit ağırlık seçen öğrencilerin %20'si, sözel seçen öğrencilerin ise %5'i ingilizceyi sevmektedir. Bu durumda ingilizce seven herhangi bir kişinin sayısal bölümde olma olasılığı bulunmak istendiğinde Bayes Teoremi kullanılabilir.

## Q5

### Normal Dağılım

1- Simetriktir.

2- Asimptotiktir.

3- (-∞, +∞) aralığında değerler alır.

4- Eğri altındaki toplam alanın olasılığı 1'dir. 

5- Ortalama mod medyan değerleri eşittir.



### Poisson Dağılımı

1- Poisson dağılımı gösteren rassal bir değişken için beklenen değer ve varyans değeri de λ'dır.

2- Tam sayılı olmayan bir λ lambda parametreli Poisson dağılımı gösteren bir rassal değişkenin mod değeri, λ 'dan küçük olan en büyük pozitif tam sayıya, yani |λ|'ya eşittir.

3- Poisson dağılımı için tüm kümülantlar beklenen değer olan λ'ya eşittirler.

4- Poisson dağılımları sonsuz olarak bölünebilir olasılık dağılımlarıdır.



### Gamma Dağılımı

1- Gamma dağılımı sonsuz bölünebilirlik özelliği gösterir.

2- Gamma dağılımı iki-parametreli üstel ailesinin bir üyesidir.

3- Üstel ve ki-kare dağılımlarının tanımlanmasına yarar.

4- Dağılımın parametreleri α ve λ'dır.

5- λ, Poisson dağılımında olduğu gibi birim aralıkta meydana gelen ortalama olay sayısını ifade etmektedir.



### Weibull Dağılımı

1- Güvenilirlik analizinde kullanılan önemli bir olasılık dağılımıdır.

2- Fonksiyonun parametreleri α ve ß'dır.

3- α=1 olursa Weibull dağılımı üstel dağılıma dönüşür.

4- α büyüdükçe dağılım normal dağılıma yaklaşır.



### Logaritmik Dağılım

1- Bir değişimin logaritmasının normal dağılım göstermesi durumudur.

2- Dağılımın parametleri λ ve ζ'dir.

3- Değişkeni daima pozitif değer alır.


